[Range slider] better sanitization and tests

danrosenthal opened the pull request:
- ~moves `value` out of `mockProps` to avoid casting as `[number, number]`~
- partially addresses #841
- handles cases where the lowerValue and upperValue are out of order
- handles cases where the lowerValue is higher than the max
- handles cases where the upperValue is lower than the min
- tests all sanitization cases

solonaarmstrong commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -20,15 +19,15 @@ describe('<DualThumb />', () => {
     it('is used to set the id on the wrapper div', () => {
       const dualThumb = mountWithAppProvider(
-        <DualThumb {...mockProps} id="RangeSlider" />,
+        <DualThumb {...mockProps} value={[0, 1]} />,
```
If this is testing the `id`, I don't understand why `id="RangeSlider"` was removed here?

danrosenthal replied:
It is already accounted for in `mockProps`, so was unnecessary to override here. Also whatever we are passing in as `mockProps` would wipe out conflicting values passed as props. Unless we are passing something different from what we passed to mockProps, it's fine just to use whatever we used there.

solonaarmstrong commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -47,112 +46,104 @@ describe('<DualThumb />', () => {
   describe('min', () => {
     it('is used to set the aria-valuemin on the lower thumb', () => {
-      const min = 0;
       const dualThumb = mountWithAppProvider(
```
Similar comment here. I don't understand why `min` was removed.

danrosenthal replied:
Ditto my previous comment.

danrosenthal commented:
@solonaarmstrong I've addressed your feedback, as well as deleted a couple of tests that are now being duplicated by my sanitization tests. Only did this in the interest of a nice separation of concerns.

danrosenthal commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -1,13 +1,12 @@
-  const mockProps = {
+  const mockProps: DualThumbProps = {
     id: 'RangeSlider',
-    cssVarPrefix: '--Polaris-RangeSlider-',
```
This was no longer being used

danrosenthal commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -19,119 +18,88 @@ describe('<DualThumb />', () => {
-      const value = [0, 20] as [number, number];
-      const dualThumb = mountWithAppProvider(
-        <DualThumb {...mockProps} min={min} value={value} />,
-      );
```
Deleted as this is tested for in the new sanitization tests.

danrosenthal commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -19,119 +18,88 @@ describe('<DualThumb />', () => {
-      const thumbUpper = findByTestID(dualThumb, 'thumbUpper');
-      expect(thumbUpper.prop('aria-valuenow')).toBe(max);
-    });
-  });
```
Deleted as this is tested for in the new sanitization tests.

danrosenthal commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -19,119 +18,88 @@ describe('<DualThumb />', () => {
-
-      const thumbLower = findByTestID(dualThumb, 'thumbLower');
-      expect(thumbLower.prop('aria-valuenow')).toBe(step);
-    });
```
Deleted as this is tested for in the new sanitization tests.

solonaarmstrong commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -19,119 +18,88 @@ describe('<DualThumb />', () => {
   describe('id', () => {
     it('is used to set the id on the wrapper div', () => {
+      const testID = 'MytestID';
+      const {id, ...rest} = mockProps;
```
Neat idea

solonaarmstrong commented on src/components/RangeSlider/components/DualThumb/DualThumb.tsx:
```diff
@@ -14,16 +14,18 @@ import {Key} from '../../../../types';
 
 import * as styles from './DualThumb.scss';
 
+export type DualValue = [number, number];
```
much better!

solonaarmstrong commented on src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx:
```diff
@@ -258,6 +258,24 @@ describe('<DualThumb />', () => {
     });
   });
 
+  describe('onChage()', () => {
```
`onChange`

danrosenthal replied:
😅

solonaarmstrong commented on src/components/RangeSlider/components/DualThumb/DualThumb.tsx:
```diff
@@ -39,17 +39,24 @@ const OUTPUT_TIP_SIZE = 8;
     const {prevValue} = state;
 
     if (isEqual(prevValue, value)) {
       return null;
```
I still think this might be fine as just `return;`

solonaarmstrong commented:
The latest commits work for me, although swapping values isn't perfect. If it's not the right approach, we can revert my commits.

danrosenthal commented:
Going to go ahead and ship this as-is. I feel good about these improvements
